### PR TITLE
GeoTiffRasterSource reproject and resample should buffer the source pixels extent

### PR DESCRIPTION
**Pull request by pomadchin**
# Overview

This PR fixes the TargetRasterExtent in GeoTiffRasterSources to represent the center of pixels.

## Checklist

- [x] [./CHANGELOG.md](https://github.com/locationtech/geotrellis/blob/master/CHANGELOG.md) updated, if necessary. Link to the issue if closed, otherwise the PR.

Closes https://github.com/locationtech/geotrellis/issues/3267

### CODE REVIEW

File: raster/src/main/scala/geotrellis/raster/geotiff/GeoTiffReprojectRasterSource.scala [hunk 1]
```diff
@@ -112,22 +112,31 @@ class GeoTiffReprojectRasterSource(
...
       val targetRasterExtent = RasterExtent(
-        extent = gridExtent.extentFor(targetPixelBounds, clamp = true),
+        extent = targetExtent,
+        cols = targetPixelBounds.width.toInt,
+        rows = targetPixelBounds.height.toInt
+      )
+
+      // buffer the targetExtent to read a buffered area from the source tiff
+      // so the resample would behave properly on borders
+      val bufferedTargetRasterExtent = RasterExtent(
```
Comment (echeipesh) [inline raster/src/main/scala/geotrellis/raster/geotiff/GeoTiffReprojectRasterSource.scala:124]: This is RasterExtent for no reason? You're just buffering the extent and then back-projecting that. Just a little confusing.

File: raster/src/main/scala/geotrellis/raster/geotiff/GeoTiffReprojectRasterSource.scala [hunk 2]
```diff
@@ -112,22 +112,31 @@ class GeoTiffReprojectRasterSource(
...
-      val sourcePixelBounds = closestTiffOverview.rasterExtent.gridBoundsFor(sourceExtent, clamp = true)
+      val sourceExtent = Proj4Transform.synchronized(bufferedTargetRasterExtent.extent.reprojectAsPolygon(backTransform, 0.001).getEnvelopeInternal)
+      val sourcePixelBounds = closestTiffOverview.rasterExtent.gridBoundsFor(sourceExtent)
       (sourcePixelBounds, targetRasterExtent)
     }}.toMap
 
     geoTiffTile.crop(intersectingWindows.keys.toSeq, bands.toArray).map { case (sourcePixelBounds, tile) =>
       val targetRasterExtent = intersectingWindows(sourcePixelBounds)
-      val sourceRaster = Raster(tile, closestTiffOverview.rasterExtent.extentFor(sourcePixelBounds, clamp = true))
+      val sourceRaster = Raster(tile, closestTiffOverview.rasterExtent.extentFor(sourcePixelBounds))
```
Comment (echeipesh) [inline raster/src/main/scala/geotrellis/raster/geotiff/GeoTiffReprojectRasterSource.scala:139]: Was removing clamp=true here critical to the fix? I'm having trouble thinking through the implications of doing this so curious to verify.

File: store/src/main/scala/geotrellis/store/cog/package.scala [hunk 3]
```diff
@@ -32,13 +32,8 @@ package object cog extends Implicits {
     }
   }
 
-  implicit class withExtentMethods(extent: Extent) {
+  implicit class ExtentLayoutOps(extent: Extent) {
```
Comment (echeipesh) [inline store/src/main/scala/geotrellis/store/cog/package.scala:35]: should avoid changing class names for little reason, saves a binary comparability bug for somebody out there in the universe.

File: raster/src/main/scala/geotrellis/raster/geotiff/GeoTiffReprojectRasterSource.scala (hunk 2 above)
Comment (echeipesh) [reply raster/src/main/scala/geotrellis/raster/geotiff/GeoTiffReprojectRasterSource.scala:139]: oh, the default value of clamp is true, right.

Comment (pomadchin) [reply raster/src/main/scala/geotrellis/raster/geotiff/GeoTiffReprojectRasterSource.scala:139]: @echeipesh it is not, it is just confusing for me to read clamp=true and this is a default behavior. I can roll it back.

**Pull request merged**